OKTA-488588 - New User query guide draft (#3382)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/reference/user-query/index.md

---
title: User query options
meta:
  - name: description
    content: Learn how to use the search parameter for most user query operations.
---

# User query options

Searching for and returning Okta users is a standard Users API lifecycle operation. The Users API supports four options to return an individual, subset, or all users:

> - [search parameter](#search-users): Returns one or more users matched against a search expression and user object properties. 

Reviewer: ```suggestion
- [search parameter](#search-users): Returns one or more users matched against a search expression and User object properties
```

- [filter parameter](#filter-users): Returns one or more users that match a filter expression checked against a subset of user object properties.

> - [find parameter](#find-users): Returns one or more users matched against the user profile properties of `firstName`, `lastName`, or `email`.

Reviewer: ```suggestion
- [find parameter](#find-users): Returns one or more users matched against the user profile properties of `firstName`, `lastName`, or `email`
```

- [list-all-users](#list-all-users): Returns all users that don't have a `DEPROVISIONED` status 

> **Note:** Okta recommends using the `search` parameter when querying for users.

For details on user operations, see the [Users API](/docs/reference/api/users/) reference.

## Search users

The search query parameter (`search`) returns one or more users matched against a search expression and User object properties, including:

- Any user profile property, including custom-defined properties
- The top-level body properties `id`, `status`, `created`, `activated`, `statusChanged`, and `lastUpdated`
- The User Type accessed from `type.id`
- Properties that have array values

This query parameter provides the largest range of search options and optimal performance. The `search` parameter requires URL encoding for applicable search expressions, supports pagination, and accepts sorting parameters.

> The search query parameter uses standard Okta API filtering semantics to create search criteria that includes mathematical operators such as equal to (`eq`), greater than or equal to (`ge`), and so on. Multiple expressions can be combined using logical operators and parentheses. See [Filtering](/docs/reference/core-okta-api/#filter).

Reviewer: ```suggestion
The search query parameter uses standard Okta API filtering semantics to create search criteria that includes mathematical operators such as equal to (`eq`), greater than or equal to (`ge`), and so on. You can combine multiple expressions using logical operators and parentheses. See [Filtering](/docs/reference/core-okta-api/#filter).
```

> The search query parameter uses standard Okta API filtering semantics to create search criteria that includes mathematical operators such as equal to (`eq`), greater than or equal to (`ge`), and so on. Multiple expressions can be combined using logical operators and parentheses. See [Filtering](/docs/reference/core-okta-api/#filter).

Reviewer: ```suggestion
The search query parameter uses standard Okta API filtering semantics to create search criteria that includes mathematical operators such as equal to (`eq`), greater than or equal to (`ge`), and so on. You can combine multiple expressions using logical operators and parentheses. See [Filtering](/docs/reference/core-okta-api/#filter).
```

        "status": "PROVISIONED",
        "created": "2021-12-17T15:16:55.000Z",
        "activated": null,
        "statusChanged": null,
        "lastLogin": null,
        "lastUpdated": "2022-05-24T15:38:04.000Z",
        "passwordChanged": "2021-12-17T15:16:56.000Z",
        "type": {
            "id": "oty2di92xFgQTiKEJ1d6"
        },
        "profile": {
            "firstName": "Janice",
            "lastName": "Benson",
            "mobilePhone": null,
            "secondEmail": null,
            "department": "Engineering",
            "login": "janice.benson@gmail.com",
            "email": "janice.benson@gmail.com"
        },
     }
```

### Limits and pagination

You can include the parameters `limit` and `after` in search expressions to limit the return of user records and to access the [pagination](/docs/reference/core-okta-api/#pagination) cursor location. By using the `after` parameter with the `limit` parameter, you can define the cursor location in the data set and manage the user records per page. The cursor is an opaque value that is obtained through the link header [next link relation](/docs/reference/core-okta-api/#link-header).

> > **Note:** If you don't specify a value for limit, the maximum (200) is used as a default.

Reviewer: ```suggestion
> **Note:** If you don't specify a value for `limit`, the maximum (200) is used as a default.
```

#### Request example

> The following example returns users who have a status of ACTIVE and limits the records returned to 5.

Reviewer: ```suggestion
The following example returns users who have an `ACTIVE` status and limits the records returned to five.
```

```bash
curl -v -X GET \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://${yourOktaDomain}/api/v1/users?search=status+eq+%22ACTIVE%22&limit=5"
```

#### Response header
The request returned the first five users and included the following next link relation in the header:

```JSON
HTTP/1.1 200 OK
Content-Type: application/json
Link: <https://{yourOktaDomain}/api/v1/users?search=status+eq+%22ACTIVE%22&limit=5>; rel="self"
Link:<https://{yourOktaDomain}/api/v1/users?after=00u3p62n11UKppeh31d7&limit=5&search=status+eq+%22ACTIVE%22>; rel="next"
```

#### Request example

The following example returns the next 5 users with the status of ACTIVE using the next link relation:

```bash
curl -v -X GET \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://${yourOktaDomain}/api/v1/users?after=00u3p62n11UKppeh31d7&limit=5&search=status+eq+%22ACTIVE%22"
```

### Sorting parameters

> The sorting parameters `sortBy` and `sortOrder` can be included in search expressions to deliver results sorted by any single user property, for example, returning users sorted by department or employee number. All users returned by a `sortBy` expression are, by default, ordered by their user `id`. Use `sortBy` to specify the property you want to sort by, and `sortOrder` to specify the order (asc and desc). By default, the sort order is ascending.

Reviewer: ```suggestion
You can include the sorting parameters `sortBy` and `sortOrder` in search expressions to deliver results sorted by any single user property, for example, returning users sorted by department or employee number. All users returned by a `sortBy` expression are, by default, ordered by their user `id`. Use `sortBy` to specify the property that you want to sort by, and `sortOrder` to specify the order (`asc` and `desc`). By default, the sort order is ascending.
```

> You can include the sorting parameters `sortBy` and `sortOrder` in search expressions to deliver results sorted by any single user property, for example, returning users sorted by department or employee number. All users returned by a `sortBy` expression are, by default, ordered by their user `id`. Use `sortBy` to specify the property that you want to sort by, and `sortOrder` to specify the order (`asc` or `desc`). By default, the sort order is ascending.

Reviewer: ```suggestion
You can include the sorting parameters `sortBy` and `sortOrder` in search expressions to deliver results sorted by any single user property, for example, returning users sorted by department or employee number. All users returned by a `sortBy` expression are, by default, ordered by their user `id`. Use `sortBy` to specify the property that you want to sort by and `sortOrder` to specify the order (`asc` or `desc`). By default, the sort order is ascending.
```

> You can include the sorting parameters `sortBy` and `sortOrder` in search expressions to deliver results sorted by any single user property, for example, returning users sorted by department or employee number. All users returned by a `sortBy` expression are, by default, ordered by their user `id`. Use `sortBy` to specify the property that you want to sort by, and `sortOrder` to specify the order (`asc` and `desc`). By default, the sort order is ascending.

Reviewer 2: Should (`asc` and `desc`) instead be (`asc` or `desc`)?  It's not like a request can specify both!

Author: Very true. Updated. Thanks.

#### Request example

The following example sorts all users with a status of DEPROVISIONED by last name using the default sort order (ascending):

```bash
curl -v -X GET \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://${yourOktaDomain}/api/v1/users?search=status+eq+%22DEPROVISIONED%22&sortBy=profile.lastName"
```

### Array searches

> The `search` parameter can also search array data types for custom user profile properties. Okta follows the [SCIM Protocol Specification]9https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.2.2) for searching arrays. If any element of the array matches the search term, the entire array (object) is returned.

Reviewer: ```suggestion
The `search` parameter can also search array data types for custom user profile properties. Okta follows the [SCIM Protocol Specification](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.2.2) for searching arrays. If any element of the array matches the search term, the entire array (object) is returned.
```

```

#### Response example

For brevity, only the user information and custom profile attribute are shown.

```JSON
"id": "00u3q8ta4i7sbzIQv1d7",
        "status": "ACTIVE",
        "created": "2022-05-20T14:41:39.000Z",
        "activated": "2022-05-20T14:41:41.000Z",
        "statusChanged": "2022-05-20T14:41:41.000Z",
        "lastLogin": null,
        "lastUpdated": "2022-05-20T14:45:47.000Z",
        "passwordChanged": "2022-05-20T14:41:40.000Z",
        "type": {
            "id": "oty2di92xFgQTiKEJ1d6"
        },
        "profile": {
            "firstName": "Sylvia A.",
            "lastName": "Ray",
            "mobilePhone": null,
            "customProp1": [
                "a",
                "b",
                "c",
                "d",
                "e"
            ],

```

#### Request example

The following example returns all users who have a custom array value (`customProp1`) that includes at least one value equal to `a` or a custom array value (`customProp2`) that includes at least one value equal to `6`:

```bash
curl -v -X GET \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \

> "https://${yourOktaDomain}/api/v1/users?search=profile.customProp1+eq+%22a%22+or+profile.customProp2+eq+%226%22"

Reviewer 2: nit: If `customProp2` is defined to be an array of Numbers or Integers instead of an array of Strings, I believe the query would not need the quotes around `"6"`, and indeed might not match if they are present (I'm not sure about that).  No need to call this out, but illustrating it in passing might be nice.  E.g., instead of two `custom array value`s you could describe them as `a custom string array value` and `a custom number array value`, then omit the quotes around `6` in the query, and leave it at that?

            "id": "oty2di92xFgQTiKEJ1d6"
        },
        "profile": {
            "firstName": "Sylvia A.",
            "lastName": "Ray",
            "mobilePhone": null,
            "customProp1": [
                "a",
                "b",
                "c",
                "d",
                "e"
            ],
            "customProp2": [
                1,
                2,
                3,
                4,
                5
            ],
        }
},
{
        "id": "00u3q8uarelmiiw0H1d7",
        "status": "ACTIVE",
        "created": "2022-05-20T14:42:13.000Z",
        "activated": "2022-05-20T14:42:14.000Z",
        "statusChanged": "2022-05-20T14:42:14.000Z",
        "lastLogin": null,
        "lastUpdated": "2022-05-20T14:46:54.000Z",
        "passwordChanged": "2022-05-20T14:42:14.000Z",
        "type": {
            "id": "oty2di92xFgQTiKEJ1d6"
        },
        "profile": {
            "firstName": "Bruce A.",
            "lastName": "Ray",
            "mobilePhone": null,
            "customProp1": [
                "f",
                "g",
                "h",
                "i",
                "j"
            ],
            "customProp2": [

>                 6,

Reviewer 2: Oh, and this really eats at me, so could you please use `7` instead of `6` in the example?  The thing that eats at me is that _every single one_ of your examples of using `eq` on an array turns out to match the _first_ element of the array.  Can we _please_ include at least one example where the matching element is _not_ the first?

Author: Done. Thanks.

        "type": {
            "id": "oty2di92xFgQTiKEJ1d6"
        },
        "profile": {
            "firstName": "Bruce A.",
            "lastName": "Ray",
            "mobilePhone": null,
            "customProp1": [
                "f",
                "g",
                "h",
                "i",
                "j"
            ],
            "customProp2": [
                6,
                7,
                8,
                9,
                10],
        }
}

```

For further search expression examples and reference material, see [Find Users](/docs/reference/api/users/#find-users) in the Users API reference.

## Filter users

The filter query parameter (`filter`) returns one or more users that match a filter expression checked against the following subset of user object properties: `status`, `lastUpdated`, `id`, `profile.login`, `profile.email`, `profile.firstName`, and `profile.lastName`. The filter query parameter supports [pagination](#limits-and-pagination) and requires [URL encoding](#url-encoding) for applicable filter expressions.

> **Note:** For optimal performance, Okta recommends using a `search` parameter instead. See [Search users](#search-users).

> The filter query parameter also uses standard Okta API filtering semantics to create filtering criteria similar to the search parameter. Multiple expressions can be combined using logical operators and parentheses. See [Filtering](/docs/reference/core-okta-api/#filter). The filter query parameter also supports the `limit` and `after` parameters, see [Limits and pagination](#limits-and-pagination).

Reviewer: ```suggestion
The filter query parameter also uses standard Okta API filtering semantics to create filtering criteria similar to the search parameter. You can combine multiple expressions using logical operators and parentheses. See [Filtering](/docs/reference/core-okta-api/#filter). The filter query parameter also supports the `limit` and `after` parameters.  See [Limits and pagination](#limits-and-pagination).
```

> The filter query parameter only uses the equal (`eq`) operator of the standard Okta API filtering semantics. The `lastUpdated` property, however, can also implement the inequality operators greater than (`gt`), greater than or equal to (`ge`), less than (`lt`), and less than or equal to (`le`). For example, you can use these operators to filter users updated after or before a specific date and time. Multiple expressions can be combined using the logical operators `and` and `or`, as well as parentheses. See [List users with a filter](/docs/reference/api/users/#list-users-with-a-filter) for example expressions.

Reviewer: ```suggestion
The filter query parameter only uses the equal (`eq`) operator of the standard Okta API filtering semantics. The `lastUpdated` property, however, can also implement the inequality operators greater than (`gt`), greater than or equal to (`ge`), less than (`lt`), and less than or equal to (`le`). For example, you can use these operators to filter users updated after or before a specific date and time. You can combine multiple expressions using the logical operators `and` and `or`, as well as parentheses. See [List users with a filter](/docs/reference/api/users/#list-users-with-a-filter) for example expressions.
```

            }
        },
        "_links": {
            "self": {
                "href": "https://{yourOktaDomain}/api/v1/users/00u17p1t4q3RNcjFc1d7"
            }
        }
    }

```

For further filter expression examples and reference material, see [Find Users](/docs/reference/api/users/#find-users) in the Users API reference.

## Find users

The find users query parameter (`q`) returns one or more users matched against the user profile properties of `firstName`, `lastName`, or `email`, and is designed for simple lookup implementations, such as a people picker. This query parameter excludes users with the status of DEPROVISIONED, does not support pagination, and is not designed for large data sets.

> **Note:** For optimal performance, Okta recommends using a [search](#search-users) parameter instead. See [Search users](#search-users).

The `q` parameter checks the prefix of the profile property (`startWith` method) to find all matches against any of the three profile properties and is not case-sensitive. The find query parameter also supports the `limit` parameter but not the `after` parameter (pagination), see [Limits and pagination](#limits-and-pagination).

> > **Note:** If you are using the `q` parameter, the default limit is 10 users if not specified.

Reviewer: ```suggestion
> **Note:** If you're using the `q` parameter, the default limit is 10 users if not specified.
```

> The `q` parameter checks the prefix of the profile property (`startWith` method) to find all matches against any of the three profile properties and isn't case-sensitive. The find query parameter also supports the `limit` parameter but not the `after` parameter (pagination), see [Limits and pagination](#limits-and-pagination).

Reviewer 2: nit: Second sentence is a run-on; use semi-colon or parentheses instead of the comma.

Author: I assume you meant the first sentence. I recast as:

`The `q` parameter checks the prefix of the profile property (`startWith` method) to find all matches against any of the three profile properties. The matching method isn't case-sensitive.`

Reviewer 2: Actually I meant the second sentence, which read: "Blah also supposes foo but not bar, see baz."  The final ", see baz" is a run-on.

Author: Yes, definitely clunky with all those parameters in there. How about:

The find users query can use limits (`limits`), but doesn't support pagination (`after`). See [Limits and pagination](#limits-and-pagination).

> The `q` parameter checks the prefix of the profile property (`startWith` method) to find all matches against any of the three profile properties. The matching method isn't case-sensitive. The find users query can use limits (`limit`) but doesn't support pagination (`after`). See [Limits and pagination](#limits-and-pagination).

Reviewer: ```suggestion
The `q` parameter checks the prefix of the profile property (`startWith` method) to find all matches against any of the three profile properties. The matching method isn't case-sensitive. The find users query can use limits (`limit`), but doesn't support pagination (`after`). See [Limits and pagination](#limits-and-pagination).
```

> > **Note:** If you're using the `q` parameter, the default limit is 10 users if not specified.

Reviewer 2: Doesn't "default" imply "if not specified"?  (Keeping the redundancy may still be right; I'm just pointing it out.)

Author: I'll remove. It's redundant to me.

        "statusChanged": "2022-05-13T20:06:22.000Z",
        "lastLogin": null,
        "lastUpdated": "2022-05-13T20:06:22.000Z",
        "passwordChanged": "2022-05-13T20:06:21.000Z",
        "type": {
            "id": "oty2di92xFgQTiKEJ1d6"
        },
        "profile": {
            "firstName": "Jack",
            "lastName": "Phillips",
            "mobilePhone": null,
            "secondEmail": null,
            "login": "john.j.phillips@example.com",
            "email": "john.j.phillips@example.com"
        },
        "credentials": {
            "password": {},
            "emails": [
                {
                    "value": "john.j.phillips@example.com",
                    "status": "VERIFIED",
                    "type": "PRIMARY"
                }
            ],
            "provider": {
                "type": "OKTA",
                "name": "OKTA"
            }
        },
        "_links": {
            "self": {
                "href": "https://{yourOktaDomain}/api/v1/users/00u3ojdzgjrAhuYGg1d7"
            }
        }
    }
```

See [Find Users](/docs/reference/api/users/#find-users) in the Users API reference.

## List all users

> The basic list-all-users command returns all users that do not have the status of DEPROVISIONED to the default maximum limit of 200 records, if not specified. For this simple use-case, the list-all query performance is superior to other search options.

Reviewer: ```suggestion
The basic list-all-users command returns all users, that don't have a `DEPROVISIONED` status, to the default maximum limit of 200 records, if not specified. For this simple use case, the list-all query performance is superior to other search options.
```

> **Note:** For most scenarios, use a [search](#search-users) parameter to refine the number of users returned.

> The list all users command supports [limits and pagination](#limits-and-pagination) but not sorting.

Reviewer: ```suggestion
The list all users command supports [limits and pagination](#limits-and-pagination), but not sorting.
```

> The basic list-all-users command returns all users, that don't have a `DEPROVISIONED` status, to the default maximum limit of 200 records, if not specified. For this simple use case, the list-all query performance is superior to other search options.

Reviewer 2: This feels redundant ("list all users command returns all users") and the hyphens feel clunky.  Do we really need to give this its own name, or could we say, "If a query does not include any of the parameters `search`, `filter`, or `q`, it returns all users (except those with `DEPROVISIONED` status), with a default limit of 200 records."

Author: Thanks. Reads much better.

> If a query does not include any of the parameters `search`, `filter`, or `q`, it returns all users (except those with DEPROVISIONED status), with a default limit of 200 records. For this simple use case, the list-all query performance is superior to other search options.

Reviewer: ```suggestion
If a query doesn't include the `search`, `filter`, or `q` parameter, it returns all users (except those with a `DEPROVISIONED` status), with a default limit of 200 records. For this simple use case, the list-all query performance is superior to other search options.
```
